### DISCUSSION TITLE
Suggestion: by default put `Save As:` in the context menu

### DISCUSSION THREAD

**Discussion by Raytle** [hluk/CopyQ#1710] [Ideas]
I don't need this feature, but I think—by default—`Save As:` should be in the context menu.

I don't need this feature because I just added `Save As:` to my context menu (the menu that pops up when I right mouse click on an item in CopyQ) by following these directions https://github.com/hluk/CopyQ/issues/252

```
[Command]
Command="
    copyq:
    tab(selectedtab())
    var item = getitem(currentitem())
    var formats = ['text/plain']

    var suffices = {
      'image/svg': 'svg',
      'image/png': 'png',
      'image/jpeg': 'jpg',
      'image/jpg': 'jpg',
      'image/bmp': 'bmp',
      'text/html': 'html',
      'text/plain' : 'txt',
    }

    function addSuffix(fileName, format)
    {
      var suffix = suffices[format]
      if ( suffix && !new RegExp(\"\\\\.\" + suffix + \"$\").test(fileName) )
        return fileName + \".\" + suffix
      return fileName
    }

    for (var format in item) {
      if (!/^application\\/x/.test(format))
        formats.push(format)
    }

    currentpath(Dir().homePath())
    var defaultFileName = 'untitled'

    var keyFormat = 'Format'
    var keyFileName = 'File'

    var result = dialog(
      '.title', 'Save Item As...',
      '.width', 250,
      keyFormat, formats,
      keyFileName, File(defaultFileName)
    )

    var fileName = result && result[keyFileName]
    var format = result && result[keyFormat]

    if (!format || !fileName)
      abort()

    fileName = addSuffix(fileName, format)

    var f = File(fileName)
    if (!f.open()) {
      popup('Failed to open \"' + f.fileName() + '\"', f.errorString())
      abort()
    }

    f.write(item[format])
    popup(\"Item Saved\", 'Item saved as \"' + f.fileName() + '\".')"
Icon=\xf0c7
InMenu=true
Name=Save Item As...
Shortcut=Ctrl+S
```

I would guess that uses who use the right mouse click (the context menu) will likely to want to `Save As` occasionally.

**Comment by hluk** [comment]
I would like to keep the core application as simple as possible and make it extendable through commands rather than having many features in the app from start. For me, it is much easier and faster to fix issues in commands.

**Comment by Raytle** [comment]
If you don't want to add it by default then it should be a feature users can easily add by merely checking a box.

> I would like to keep the core application as simple as possible and make it extendable through commands rather than having many features in the app from start. 

I'll be blunt and, well, even **a little harsh**: that's a laudable goal but misses the point... almost entirely. You are making a common UI/UX design error. And let's be clear: it is a design error. Dictating to users how they must use an application unless they want to "wade into the deep end of the pool" to learn arcane commands is, well, it's something developers often require. But based on [Design for the Novice, Configure for the Pro](https://bothsidesofthetable.com/design-for-the-novice-configure-for-the-pro-b259c6b4662a) it's bad practice Very bad practice.

> For me, it is much easier and faster to fix issues in commands.

That's irrelevant to this discussion, and, well it's also self-centered. You obviously can create a GUI element that would enable users to toggle a feature such as "Save As." If you don't want to bother, well, then that's your prerogative. After all, you are the maintainer of this FOSS project. You are free to ignore valid criticisms and good advice... if you would like.

Here's an example. The "Feature-Packed Text Editor" [Kate](https://kate-editor.org/) has myriad user configurable options.

![Image 1](https://user-images.githubusercontent.com/81956552/126415256-9d4e1433-e8dd-45d6-8099-79585a6b877a.png)

Seems like chaos. Right?

When I use Kate, except for "Show Scrollbar Mini-Map" (which I pointed to with a green arrow) I normally have, no other Kate GUI elements visible, not even drop down-menus.

![Image 1](https://user-images.githubusercontent.com/81956552/126415541-5151b301-12e5-4404-b99c-601d311c549e.png)

**Reply by hluk** [reply]
Those are my design decisions. I do not have time to maintain many features in multi-platform application - even in current state it is not easy for me.

I would rather let users to extend the app themselves and make sharing commands easier.

**Reply by Raytle** [reply]
> I would rather let users to extend the app themselves and make sharing commands easier.

I agree 100% with that. My suggestions dovetail (fit) nicely with that philosophy. I think you might have misunderstood my suggestion.


> Those are my design decisions. 

They are **bad** design decisions.


> I do not have time to maintain many features in multi-platform application

Bringing up "multi-platform application" seems to be a [red herring](https://en.wiktionary.org/wiki/red_herring#Noun). 

`(figuratively) A clue, information, argument etc. that is or is intended to be misleading, diverting attention from the real answer or issue.)`

I'm talking about a design philosophy (in other words, "functional architecture") which has nothing to do with Windows vs Linux vs Mac.


 > even in current state it is not easy for me.

I have noticed that you spend **huge amounts of time** helping users by answering their questions. That time is being **squandered**. Why do I assert that? With proper design users would be able to solve many of their problems themselves and therefore not need to ask you questions in the first place.

Look at "Save As" for example. If "Save As" in CopyQ were included in an interface similar to this...

![trash](https://user-images.githubusercontent.com/81956552/126576341-4f972c33-f5c8-4bd6-82c0-b122ba2d22b7.png)

then ordinary users could easily create a shortcut for "Save As." 

Requiring **engineers** to go to, for example, `File –> Commands/Global Shortcuts...` and then create a new command is not unreasonable; however, requiring **ordinary users** to do that is utterly, totally, and completely ridiculous. It's terrible design. 

Put simply: bad design tends to result in users asking questions that would have been unnecessary if good design had been used in the first place.

**Comment by Raytle** [comment]
Are you familiar with the [Recoll](https://www.lesbonscomptes.com/recoll/)? It's a desktop search tool which I use frequently. A few months ago I helped the developer to create this...

![abc](https://user-images.githubusercontent.com/81956552/126577184-8091fa88-d49b-431d-824b-a2c37670dc7f.png)

Don't you see how easy it is for users to customize? It's an example of good user interface design.

Do you notice I have only customized one setting? To me `ctrl + f` makes more sense than `ctrl + s` because I think of "find" not "search". But here's the thing: that makes Recoll more "user friendly" for me to use. Why? Because when I am looking for something in a Google Doc or a text document in Kate or Firefox I normally press `ctrl + f`. And guess what? In Recoll when I am looking for something  I can also press `ctrl + f`.

Each of these little things is not very significant. Sure. I know that. But in total (in sum) they are significant. Look at how you have designed your programming environment. If you are like most developers you have spent countless hours customizing it in a way that is easy for you to use. Good design is important; it is very, very, very important.

An small, insignificant, and obscure technology company with a fruit in its name—yes, a fruit!—figured this out decades ago. And no. Apple Inc is no longer small, insignificant, and obscure. At Apple they understand that user interface design and user experience design are crucial.
